Support Widget display on Categories Page and Fix Text Color using textContrastClass

Yippy opened the pull request:
Added new settings to enable or disable different sections of widgets
![Categories Page Support for widgets](https://github.com/askvortsov1/flarum-categories/assets/6518915/ae06f6aa-3994-492b-a2e4-c5a2b4f206b2)


Fix #25 Using textContrastClass and applying filter for sub heading contrast
![Text Color Fix](https://github.com/askvortsov1/flarum-categories/assets/6518915/e80889aa-bef5-48d0-b170-fadadd49d1cd)

Yippy committed:
Support Widget display for categories, such as Statistics Widget, Custom Html Widget, Afrux Forum Widgets

Yippy committed:
Fix #25 Use textContrastClass to switch between white and black font color depending on the color of the tag and amend the style of the sub heading with filter brightness for different color contrast

Yippy commented:
For afrux left sidebar top and left sidebar bottom to display correctly within Categories page, the PR afrux/forum-widgets-core#17 needs to be merged. But header, footer, and right widget will display correctly without the PR.

Yippy commented:
Working on fixing dark mode, will remove the workaround as this causes issues. Will also remove filter brightness for just applying styling.

Yippy committed:
Redo dark theme support, removed declared colors and use css names to match with forum.js
Added 2 style theming for both parent and child tags

askvortsov1 commented on extend.php:
```diff
@@ -39,7 +39,13 @@
-        ->serializeToForum('categories.childBareIcon', 'askvortsov-categories.child-bare-icon', 'boolval', true),
+        ->serializeToForum('categories.childBareIcon', 'askvortsov-categories.child-bare-icon', 'boolval', true)
+        ->serializeToForum('categories.widgetHeader', 'askvortsov-categories.widget-header', 'boolval', false)
+        ->serializeToForum('categories.widgetRight', 'askvortsov-categories.widget-right', 'boolval', false)
```
Why not default these to true? If someone has installed a widget extension, presumably they want widgets?

Yippy replied:
Not entirely, I though it was better to turn them off by default with the original categories page. No one seems to have complained about not having widgets on the page. The original didn't have header, footer or right side widget. Only there was a bug for allowing the widget to show on the left sidebar, which I accidently fixed. This PR was to make it almost like before, but with options

askvortsov1 commented on js/src/forum/components/CategoriesPage.tsx:
```diff
@@ -56,18 +56,38 @@ export default class CategoriesPage extends Page {
-
+    // Only check for widget for the header and footer if enable in the settings
+    if ((app.forum.attribute('categories.widgetHeader') || app.forum.attribute('categories.widgetFooter')) && indexPage.children.length > 1 && indexPage.children[1].children) {
+      indexPage.children[1].children.forEach((child, index) => {
```
Why is this happening in a loop? Why not use `children.some()` to determine whether there exists a child that would trigger the widgets to be added? And optional chaining to simplify the if statement?

That being said, this code is very brittle: structure of vdom nodes is subject to change, and relying on the absence of a class name is prone to go wrong as soon as someone sees one.

Is there some value / better API we could expose in Core that would make this more reliable?

Yippy replied:
I will work on a better detection on this later

Yippy replied:
I haven't learnt of a way to grab the non populated extended in a super reliable way. I have made a better method in going through known classname instead and look for undefined className because they are from the widget extension. That way the extension can search through the tiers of containers inside the IndexPage.

askvortsov1 commented on js/src/forum/components/CategoriesPage.tsx:
```diff
@@ -86,15 +106,21 @@ export default class CategoriesPage extends Page {
     );
-
+    // Only check for widget for the right if enable in the settings
+    if (app.forum.attribute('categories.widgetRight')) {
```
Is there no way for the widgets plugin to provide an API for reliably getting it's widgets? Or for it to extend this plugin?

Yippy replied:
Well the plugin doesn't have option to hide or show widget on pages, I've provided a toggle to give admin the choice of showing what is needed for their categories page. Some people have them as their main page, while I don't have it as the mainpage. But the widget is useful else where on the website.

askvortsov1 commented on js/src/forum/components/Category.tsx:
```diff
@@ -141,10 +156,16 @@ export default class Category extends Component<Attrs> {
-    items.add('icon', <span className="TagCategory-icon">{this.iconItems().toArray()}</span>, 100);
+    const children = this.isChild ? [] : sortTags(this.tag.children() || []);
+    let style: Record<string, unknown> = {};
+    if (this.isChild) {
```
This "style building" code seems to be repeated; could you extract it into a helper function?

askvortsov1 commented on resources/locale/en.yml:
```diff
@@ -12,16 +12,30 @@ askvortsov-categories:
       parent_remove_stats: Hide stats for top-level tags?
       parent_remove_last_discussion: Hide most recent discussions for top-level tags?
       small_forum_optimized: Optimize for small forums?
+      enable_primary_tag_color: Enable Primary Tag Color?
```
Why don't we always want to enable tag colors?

Yippy replied:
I've shown examples of the theme, which can give admin the choice of having it all minimalistic with the dark and light support.

askvortsov1 reviewed (changes requested):
Thanks for the PR!

Yippy commented:
Theme Choices Example
![image](https://github.com/askvortsov1/flarum-categories/assets/6518915/e946a953-b6e4-479d-ae43-924b4d04d8d7)
![example of primary tag colors enabled](https://github.com/askvortsov1/flarum-categories/assets/6518915/61afd2b6-0c63-4726-912a-7465331303a1)

![image](https://github.com/askvortsov1/flarum-categories/assets/6518915/e90c0dbb-383e-45a9-bde5-69e90ba98187)
![example of primary tag colors disabled gif](https://github.com/askvortsov1/flarum-categories/assets/6518915/3572b0e1-9cfc-4e43-9c96-352e8b300113)

![image](https://github.com/askvortsov1/flarum-categories/assets/6518915/0f8f9e2a-c019-4912-b0e4-964e957c5e68)
![example of primary tag colors just parent enabled gif](https://github.com/askvortsov1/flarum-categories/assets/6518915/133041a9-2e1d-467f-8fac-ff683cb94021)

![image](https://github.com/askvortsov1/flarum-categories/assets/6518915/cbf50085-88c3-4c16-bd33-b118f239264f)
![example of primary tag colors just child enabled gif](https://github.com/askvortsov1/flarum-categories/assets/6518915/580f54c7-54f9-455c-ac05-dbc22a8ce683)

Please note that Child Tag 4 has no color assigned, it's left as '[empty]' for testing, I will work on @askvortsov1 concerns. It does need refactoring in some areas. But I'm trying to work on datlechin/flarum-tag-passwords#11 to hide protected discussion and fix the discussion protection within this first.

Yippy committed:
Fix typo and rephrase sentence

Yippy committed:
Fix ordering of ItemList

Yippy committed:
Improved widget container finder with the add findWidgetContainer function.
POSITION_ANY = No position check
POSITION_FIRST = First container only
POSITION_AFTER_FIRST = After the first container

Yippy committed:
Add new helper lastDiscussionAvatar, to support 'Anonymous' users that were last poster

Yippy committed:
Remove usage of icon, use '?' instead
